### PR TITLE
Add to DocumentFragment Usage notes of invalid HTML created with DocumentFragment

### PR DESCRIPTION
**Pull request by 0xedward**
> What was wrong/why is this fix needed? (quick summary only)

Add content to DocumentFragment Usage notes section to inform the reader that it's possible to create "invalid" HTML with `DocumentFragment` s that will be removed by the HTML parser

 
> Anything else that could help us review it

Consider the following test case:
1. Run the following code in console:
```javascript
let fragment = document.createDocumentFragment()
const bodyElement = document.createElement("body")
const trElement = document.createElement("tr")
fragment.appendChild(bodyElement)
fragment.querySelector('body').appendChild(trElement) // create invalid HTML with DocumentFragment
let testDocument = document.implementation.createHTMLDocument('test document');
testDocument.querySelector('body').replaceWith(fragment)
```
2. Check that `testDocument` is not a valid `HTMLDocument`. Here's two ways of testing this:
 - Create an HTML document (e.g. `test.html`) with the value of `testDocument`, then open `test.html` and inspect element with devtools. The `tr` element should be stripped out out of the `body` element.
```html 
<!DOCTYPE html>
<html>
  <head>
    <title>test document</title>
  </head>
  <body>
    <tr></tr>
  </body>
</html>
```
 - Paste `testDocument` into [Nu HTML Checker](https://validator.w3.org/nu/#textarea)
 
image

### CODE REVIEW

File: files/en-us/web/api/documentfragment/index.html [hunk 1]
```diff
@@ -64,6 +64,8 @@ <h2 id="Usage_notes">Usage notes</h2>
 
 <p>An empty <code>DocumentFragment</code> can be created using the {{domxref("document.createDocumentFragment()")}} method or the constructor.</p>
 
+<p><strong>Caution:</strong>It's possible to create constructs that HTML doesn't allow with <code>DocumentFragment</code>s. When those <code>DocumentFragment</code>s are inserted into a <code>Document</code> object, the HTML parser will mutate the <code>Document</code> to create a valid <code>HTMLDocument</code></p>
```
Comment (sideshowbarker) [inline files/en-us/web/api/documentfragment/index.html:67]: The HTML parser doesn’t necessarily produce valid HTML. There is never any guarantee that output from parsing will be valid. And HTML parsers don’t mutate documents. Instead they take input and build a tree from that input.

**Comment by sideshowbarker** [comment]
> Create an HTML document (e.g. `test.html`) with the value of `testDocument`, then open `test.html` and inspect element with devtools. The `tr` element should be stripped out out of the `body` element.

What that actually seems to be describing is:

1. first *serializing* `testDocument` — that is, taking the DOM of `testDocument` and generating some string representation of its elements and attributes
2. then taking that serialized representation of `testDocument` and running it through an HTML parser to re-parse it

That’s because, when you give a document to the HTML checker (validator), the checker does not check the serialized markup source — instead the checker takes that serialized markup source and runs it through an HTML parser to create a tree representation (DOM) of all the elements and attributes; and then it checks that parsed tree.

**Comment by sideshowbarker** [comment]
There is nothing about all this which is specific to document fragments.

Here’s something similar to the code in the question:

```js
const trElement = document.createElement("tr")
document.body.appendChild(trElement)
```

Now the document itself has a `tr` element child as the last child of the body. Of course the browser isn’t going to actually render anything for that, but it’s still in the document — in the DOM.

So I think the only thing that can really be said here is that in general when you’re doing direct DOM manipulation from JavaScript, it’s possible to create subtrees in the document that are essentially broken and that browsers aren’t going to render in any useful way, and that don’t follow the document-conformance requirements (authoring requirements) defined in the HTML spec.

Beyond that, I’m not sure what more we can really say. But I guess maybe one thing that could be said is that if your goal is it ensure that you’re not putting something completely broken into the DOM, you might be better off doing the document manipulation using `innerHTML`.

For example, if do this:

```js
const newBody = document.body.innerHTML + "<tr></tr>"
document.body.innerHTML = newBody
```
…then the body of the document isn’t actually going to change — because the rules for HTML parsing cause the HTML parser to just drop that `<tr></tr>` completely.

In other words, if you use `innerHTML`, you’ve at least got a better chance of preventing your code from creating junk in the DOM that doesn’t make any sense.

**Comment by 0xedward** [comment]
Oh, I see! Thanks @sideshowbarker! This was really educational :)

> The HTML parser doesn’t necessarily produce valid HTML. There is never any guarantee that output from parsing will be valid. And HTML parsers don’t mutate documents. Instead they take input and build a tree from that input.

>That’s because, when you give a document to the HTML checker (validator), the checker does not check the serialized markup source — instead the checker takes that serialized markup source and runs it through an HTML parser to create a tree representation (DOM) of all the elements and attributes; and then it checks that parsed tree.

Sounds like it's actually the browser dropping the orphan nodes rather than the HTML parser
